### DISCUSSION TITLE
`solidSubMesh` Implementation

### DISCUSSION THREAD

**Discussion by denMaier** [solids4foam/solids4foam#133] [Ideas]
## Background
I've recently encountered an issue while using `dynamicRedefineFvMesh` in one of my cases. The refinement process disrupted the cell mapping of the `solidSubMesh`, which led me to some questions.

## Key Questions
1. What is the primary rationale behind the implementation of `solidSubMesh`?
2. What specific functionality does `solidSubMesh` provide that cannot be achieved using the existing `ZoneMesh` system?

## Observations and Suggestions
1. **BiMaterial Interfaces**: I've noticed that some code relates to identifying biMaterial interfaces. However, I believe this functionality could be replicated more simply:
   - We could check the owner and neighbor cells of all faces.
   - If these cells belong to different zones, the face could be added to an interface faceZone.

2. **Simplification Potential**: From my perspective, leveraging the `ZoneMesh` functionality could offer several advantages:
   - Simplification of the codebase
   - Reduced maintenance overhead
   - Elimination of the need to store duplicate fields

3. **Current Challenges**: The current approach of maintaining separate base and subMesh fields introduces synchronization issues.

## Proposal
I propose we consider refactoring the current implementation to utilize `ZoneMesh` more extensively. This could potentially streamline our codebase and resolve some of the existing challenges, particularly those related to mesh refinement and field synchronization.

## Questions for the Community
1. Are there specific use cases or requirements that necessitated the current `solidSubMesh` implementation?
2. What potential drawbacks or challenges do you foresee in transitioning to a `ZoneMesh`-based approach?
3. Has anyone else encountered similar issues with `dynamicRedefineFvMesh` and `solidSubMesh`?

I welcome your thoughts, insights, and experiences on this matter. Let's discuss how we can improve our mesh handling to be more robust and maintainable.

**Comment by philipcardiff** [comment]
Hi @denMaier,

In short, I agree with you. The current `solidSubMesh` is overly complicated and causes various issues when interacting with other code.

There are two motivations for the existence of `solidSubMesh`, both of which are related to multi-material cases:
- The issue examined in [Tukovic et al., Finite-volume stress analysis in multi-material linear elastic body, 2012, IJNME](https://onlinelibrary.wiley.com/doi/abs/10.1002/nme.4390), where the standard cell-centred discretisation needs to be modified in cells adjacent to a bi-material interface to avoid unphysical oscillations. Essentially, when calculating the gradient in a cell by the interface, you cannot use the values in cells on the other side of the interface because the displacement gradient is discontinuous across the interface. Using `solidSubMesh` , we can calculate gradients in each region with `fvc::grad` and then perform the necessary corrections to the displacement on the bi-material interface faces. A related tutorial case is on the website: https://www.solids4foam.com/tutorials/more-tutorials/solid-mechanics/multiMaterial/layeredPipe.html.
- If we have two materials, it is "nice" if we can define volScalarFields/volVectorFields/etc. only for the cells of this material. Before the `solidSubMesh` approach, all material laws were defined for all cells, and a `multiMaterial` mechanical law class would decide which stresses to use for each cell. This can be a problem for plasticity on other material laws, since they may have problems calculating stresses in cells in the mesh, where they should not be applied.

Having said that, `solidSubMesh` was just one way to resolve the two points above and there are likely many ways, some of which are better. Now that v2.1 is out, I am open to (in fact, interested in) refactoring the mechanical law classes to make them easier to understand, maintain and extend.

I have experience with mechanical law interfaces in two other software: Abaqus and [FLAC3D](https://www.itasca.fr/en/software/flac3d) (FLAC3D is actually a finite volume solid mechanics code! albeit it is explicit and vertex-centred).

The mechanical law (they call it the constitutive model) interface for FLAC3D is:
```c++
void myMechLaw::run(UByte d,flac3dState *s);
```
where `UByte d` indicates whether the case is 2-D or 3-D (influences the storage of fields), and `flac3dState *s` is a struct the stores any material state variable, i.e. the mechanical law does not store any state variables (e.g. plastic strain, etc.)

The Abaqus explicit "VUMAT" interface is :
```fortran
– The VUMAT subroutine header is shown below: subroutine vumat(
c Read only (unmodifiable) variables -
nblock, ndir, nshr, nstatev, nfieldv, nprops, lanneal, stepTime, totalTime, dt, cmname, coordMp, charLength, props, density, strainInc, relSpinInc,
tempOld, stretchOld, defgradOld, fieldOld,
stressOld, stateOld, enerInternOld, enerInelasOld, tempNew, stretchNew, defgradNew, fieldNew,
c write only (modifiable) variables -
stressNew, stateNew, enerInternNew, enerInelasNew)
```
where `nblock` is the number of points to be calculated (integration points), and you can see there are various read-only variables specifying the dimensions (ndir, nshr), number of state variable fields, properties read in from a user file, time settings, coordinates, strain/stretch/strain-rate/deformation-gradient, energies, and finally variables to be defined like stress and new state variables. The Abaqus implicit "UMAT" interface is similar but also requires that material tangent be defined. Abaqus also provides some higher-level interfaces, e.g. for hyper elastic materials where only the derivative of the free energy wrt to the invariants must be defined. However, like FLAC3D, you will see that the material law does not store any state variables. Instead, they are passed into the function that calculates the stress. Both approaches follow a functional programming style.

Looking towards solids4foam, I could imagine us adopting a similar interface, where `Fields` are passed instead of `GeometricFields`, which would avoid the need for different functions for volFields, surfaceFields and (my new interest) pointFields. Equally, it would allow us to get rid of solidSubMeshes. We could call a material law stress calculation function on one node (Field of size 1), or a patch or the entire internalField or whatever selection of nodes we like.

@iBatistic, @scottlevie97, @ScimonCFD, @alishayegh , @cloner0110 , @federico081997 : some light reading, in case you're interested.

**Comment by denMaier** [comment]
*** Converted from replies to comments, so that the matters can be discussed seperately ***

Thanks @philipcardiff for the explanation.   

I have thoughts to both your two main points. I want to start with your second point, the one about the internal state variables of the material laws. The current approach has the advantage that the fields sit where "they belong" and are completely demand driven. In fact, the solver does not need to know which internal fields are going to be used. On the other hand they are very isolated from the rest of the code now. 
My idea: let's make use of the objectRegistry class. Every materialLaw could be its own subRegistry under the solid/region0 registry and is named after the cellZone. The internal fields register into that materialLaw. Then, one could get to any internal variables, by only knowing the name of the cellZone and doing a lookup. 
For example, one could get a list of all names of internal state variables with:
```
hashedWordList interternalStateVariableNames();
forAll(mesh().cellZones(),zoneI)
{
   word zoneName = mesh().cellZones().names()[zoneI];
   const objectRegistry& zoneRegistry = time().subRegistry("solid").subRegistry(zoneName);
   interternalStateVariableNames.append(zoneRegistry.names());
}
interternalStateVariableNames.uniq();
```

**Comment by denMaier** [comment]
*** Converted from replies to comments, so that the matters can be discussed seperately ***

To your point 1 about the use cases of solid subMeshes:
I recently restructured my `poroHydraulicModel` (eq. to `materialModel` in s4f) to use cellZones.
I decided to exclusively work on the ESI version of OpenFOAM (this is the only version IT would allow me to install..).
So i don't know if the implementation would be as straight forward in other versions. 

`cellZone` or rather the base class `ZoneMesh<type>` provides a mapping function `whichZone(label globalLabel)`
This makes it extremely easy to loop through the cells and apply the material law corresponding to the zone.

### Initialize a list of conductivityLaws, one for each cellZone
```

PtrList<conductivityModel> conductivityModelPtr_(mesh().cellZones().size());
forAll(mesh().cellZones(),zoneI)
{
        const dictionary& zoneSubDict =
             poroHydraulicProperties_.optionalSubDict 
                      (
                         mesh().cellZones().names()[zoneI]
                      );
        conductivityModelPtr_.set
        (
            zoneI,
            conductivityModel::New(...)
        );
}
``` 
### Use the created list to update each cell
``` 
void poroHydraulicModel::correctK(volScalarField& k) const
{
    const cellZoneMesh& cellZones = mesh().cellZones();  
    forAll(k,cellI)
    {
        conductivityModel& cModel = conductivityModelPtr_[cellZones.whichZone(cellI)];
        k.internalFieldRef()[cellI] = cModel.k(cellI);
    }
    forAll(k.boundaryMesh(),patchI)
    {
        scalarField& kPatch = k.boundaryFieldRef()[patchI];
        const labelUList& faceCells = mesh().boundaryField()[patchI].faceCells();
        forAll(nPatch,faceI)
        {
            label cellI = faceCells[faceI];
            conductivityModel& cModel = conductivityModelPtr_[cellZones.whichZone(cellI)];
            kPatch[faceI] = cModel.k(patchI,faceI);
        }
    }
}
``` 
Two remarks:
- I am planing to template the function instead of hardcoding the `volScalarField` argument.
- I need two functions `k(cellI)` and `k(patchI,faceI)` because I wanted the internal variables / coefficients of the functions to be `volScalarField`s, so that they can still be provided as fields in the `0` directory. Maybe there is a way to reuse a single scalar function. Maybe having access to the internal variables outside of the material law could help with that?

**Reply by denMaier** [reply]
For the bi-material interfaces:
- Iterate through all faces in the mesh
- For each face, check if it's an internal face (has both owner and neighbour cells)
- If it is, check if the owner and neighbour cells belong to different `cellZones`
- If they do, this face is an interface face -> add it to a `faceZone` named after the two cellZones it connects

We thus would get one `faceZone` for each bi-material interface. We can then correct them similar to the methodology above.

**Comment by philipcardiff** [comment]
Thanks for the interesting suggestions, @denMaier. Let me come back with some possible implementation approaches and we can hopefully iterate towards a design that ticks most boxes.
